src/libre/ast_rewrite.c

```diff
@@ -125,26 +127,22 @@ rewrite_concat(struct ast_expr *n, enum re_flags flags)
 	assert(n->flags == 0x0);
 
 	for (i = 0; i < n->u.concat.count; i++) {
-		if (!rewrite(n->u.concat.n[i], flags)) {
+		if (!rewrite(poolp, n->u.concat.n[i], flags)) {
 			return 0;
 		}
 	}
 
 	/* a tombstone here means the entire concatenation is a tombstone */
 	for (i = 0; i < n->u.concat.count; i++) {
 		if (n->u.concat.n[i]->type == AST_EXPR_TOMBSTONE) {
-			for (i = 0; i < n->u.concat.count; i++) {
-				ast_expr_free(n->u.concat.n[i]);
-			}
-
```

Comment:
Is this the double free bug?

Reply:
I tried to reproduce this with something like `./build/bin/re -r pcre -pl ast 'abc(?:$x)+def'` thinking that `(?:$x)` would become a tombstone, but that doesn't seem to do it.

Reply:
Yup, that's the double free bug. `(?:$x)` doesn't become a tombstone, that might be a useful optimization to add. `[^\x00-\xff]` and `[x^x]` (in the SQL dialect) do, so any concatenation with those will trigger it. This shows up in `tests/native/in31.re`, `tests/native/in34.re`, `tests/native/in36.re`, `tests/pcre-anchor/in52.re`, and `tests/sql/in12.re`.

Reply:
From @hvdijk:
> Ah and now I see why this didn't show up as an error before the pool: the double free was introduced by commit 9338812e, which is part of the PR to use the pool, which immediately makes it harmless :)